compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1535,22 +1580,82 @@ void GpuKernel::generateOobCond() {
+  fn_->insertAtTail(new CondStmt(new SymExpr(isInBounds), this->userBody_));
+}
+
+/* Adds the following AST to a GPU kernel
+ *
+ * // recall index[0] = lowerBound + `global thread idx` * itersPerThread
+ * def threadBound = min(index[0] + itersPerThread - 1, upperBound)
+ *
+ * CForLoop(; index[0] <= threadBound; for all i: index[i] += 1) {
+ *   // empty block stored in this->userBody_
+ * }
+ */
+void GpuKernel::generateLoopOverIPT(Symbol* upperBound) {
+  // initBlock is empty, inits are added in generateIndexComputation()
+  BlockStmt* initBlock = new BlockStmt();
+
+  // testBlock tests only for index[0], like in generateOobCondNoIPT()
+  Symbol* index0 = kernelIndices_[0];
+
+  VarSymbol* iptMinus1 = insertNewVarAndDef(fn_->body, "iptMinus1",
+                                            localItersPerThread_->type);
+  fn_->insertAtTail("'='(%S,'-'(%S,%S))", iptMinus1,
+                    localItersPerThread_, new_IntSymbol(1));
+
+  VarSymbol* threadBound = insertNewVarAndDef(fn_->body, "threadBound",
+                                              index0->type);
+  fn_->insertAtTail("'='(%S,'+'(%S,%S))", threadBound, index0, iptMinus1);
+
+  VarSymbol* switchToLB = insertNewVarAndDef(fn_->body, "switchToLB", dtBool);
+  fn_->insertAtTail("'='(%S,'<'(%S,%S))", switchToLB, upperBound, threadBound);
+
+  BlockStmt* switchBlock = new BlockStmt();
+  switchBlock->insertAtTail("'='(%S,%S)", threadBound, upperBound);
+  fn_->insertAtTail(new CondStmt(new SymExpr(switchToLB), switchBlock));
 
-  this->userBody_ = thenBlock;
+  BlockStmt* testBlock = new BlockStmt();
+  testBlock->insertAtTail("'<='(%S,%S)", index0, threadBound);
+
+  // incrBlock increments all indices
+  BlockStmt* incrBlock = new BlockStmt();
+  for_vector(Symbol, index, kernelIndices_) {
+    incrBlock->insertAtTail("'+='(%S,%S)", index, new_IntSymbol(1));
+  }
+
+  CForLoop* iptLoop = toCForLoop(
+    CForLoop::buildCForLoop(nullptr, this->userBody_) // returns a BlockStmt
+      ->body.head->remove());
+  iptLoop->loopHeaderSet(initBlock, testBlock, incrBlock);
+  fn_->insertAtTail(iptLoop);
 }
 
 void GpuKernel::generatePostBody() {
   this->postBody_ = new BlockStmt();
   fn_->insertAtTail(this->postBody_);
 }
 
-void GpuKernel::processGpuPrimitivesBlock(BlockStmt* block) {
-  INT_ASSERT(block);
-  INT_ASSERT(block->isGpuPrimitivesBlock());
+// Detect attributes if any, store them in 'this', remove them from the tree.
+void GpuKernel::processGpuPrimitivesBlock() {
+  BlockStmt* block = nullptr; // this will be the primitives block
+
+  // find the primitives block
+  // Note: this searches inside nested loops, needed for multi-dim arrays
+  // ex:  var A: [1..n,1..n] int; @gpu.blockSize foreach A {}
+  std::vector<BlockStmt*> blocksInBody;
+  collectBlockStmts(gpuLoop.loop(), blocksInBody);
+  for (auto block2 : blocksInBody) {
+    if (block2->isGpuPrimitivesBlock()) {
+      block = block2;
+      break;
+    }
+  }
+
+  if (block == nullptr) return; // did not find it
```

Comment:
consider renaming `block => primBlock` and `block2` => `block` to make intent more clear

Reply:
Right... I avoided renaming to reduce diffs. I can rename.